lib/ReactViews/DataCatalog/DisplayGroupService.ts

```diff
@@ -0,0 +1,85 @@
+import { runInAction } from "mobx";
+import { DataSourceAction } from "../../Core/AnalyticEvents/analyticEvents";
+import MappableMixin from "../../ModelMixins/MappableMixin";
+import { BaseModel } from "../../Models/Definition/Model";
+import Terria from "../../Models/Terria";
+import ViewState from "../../ReactViewModels/ViewState";
+import toggleItemOnMapFromCatalog, {
+  Op as ToggleOnMapOp
+} from "./toggleItemOnMapFromCatalog";
+
+/**
+ * Summary. A collection of functions related to displayGroups. These functions called in multiple places e.g. `GroupPreview` and `DataCatalogGroup`.
+ */
+
+/**
+ * A check to see if all members of a group are already loaded in the workbench. Only checks for mappable items.
+ */
+export function allMappableMembersInWorkbench(
+  groupItems: string[],
+  terria: Terria
+) {
+  // Check if all the mappable items from our group are already loaded in the workbench
+  const checkAllMappablesInWorkbench = (
+    workbenchArr: readonly string[],
+    groupItemsArray: string[]
+  ) =>
+    groupItemsArray.every(
+      (member: any) =>
+        !MappableMixin.isMixedInto(terria.getModelById(BaseModel, member)) ||
+        workbenchArr.includes(member)
+    );
+  return checkAllMappablesInWorkbench(terria.workbench.itemIds, groupItems);
+}
+
+/**
+ * Function to handle adding or removing of group items.
+ * Will first check to see if all remebers are currently loaded.
+ * If they are, button will remove all.
+ * If any items are missing from the workbench, it will add those that are missing, up to and including all items
+ */
+export function addRemoveButtonClicked(
+  previewedGroup: any,
+  viewState: ViewState,
+  terria: Terria,
+  keepCatalogOpen: boolean
+) {
+  runInAction(() => {
+    // Force items to be loaded or removed depending on state of entire group.
+    const forceState = !allMappableMembersInWorkbench(
+      previewedGroup.members,
+      terria
+    );
+    previewedGroup.loadMembers().then(() => {
+      for (
+        let index = previewedGroup.memberModels.length - 1;
+        index >= 0;
+        index--
+      ) {
+        const memberModel = previewedGroup.memberModels[index];
+        addOrRemoveMember(memberModel, viewState, forceState, keepCatalogOpen);
```

Comment:
Doubt: Can we avoid introducing `forceState` if we only call `addOrRemoveMember()` on a member if it is not in the state that we want it to be. i.e If user clicked "Add all" we call it only if the member is not already added to the workbench (and vice-versa).

Reply:
Yes, i think the logic I have added now should do it, good idea.

Reply:
it might be worth checking my logic here, second pair of eyes, but I think its sound

Reply:
I feel I have to pause a bit  to understand the logic. The problem could be that we only have a `toggleItemOnMapFromCatalog` to work with, so it is hard to express whether we want add or remove the item. 

One solution might be to add separate functions in the `DispalyGroupService` file for adding and removing: `addMemberToMap` and `removeMemberFromMap`. Each would internally call `toggleItemOnMapFromCatalog` after checking `viewState.terria.workbench.contains(item)`.

Then we could write something like:
```
 if (allItemsLoaded) {
          // toggle action should remove all group members from map
          removeMemberFromMap(memberModel, viewState, keepCatalogOpen);
  } else if (!itemLoaded) {
           // toggle action should add missing members to the map
          addMemberToMap(memberModel, viewState, keepCatalogOpen);
  }
```

What say @staffordsmith83 ?

Reply:
So I have fixed it with the new logic - thinking that it was best to not introduce another parameter into `toggleItemOnMapFromCatalog` function, and that it would also be best to keep using that function rather than write another one.

The logic is sound:
We check if all items are already loaded
We check if our current item of interest is already loaded
If all items are loaded, then we can just call `toggle` for all, and then they will all be removed, this is what we want.
Net we check if `!allItemsLoaded && !itemLoaded` - if we have a partially loaded group, and our current item of interest is not loaded, then we toggle it, which will load it - this is what we want.

All other cases we do nothing. I think there is only one other case not accounted for:
`!allItemsLoaded && itemLoaded` - thats fine, do nothing


I.e. like this:

```
const allItemsLoaded = allMappableMembersInWorkbench(
      previewedGroup.members,
      terria
    );
    previewedGroup.loadMembers().then(() => {
      for (
        let index = previewedGroup.memberModels.length - 1;
        index >= 0;
        index--
      ) {
        const memberModel = previewedGroup.memberModels[index];
        const itemLoaded = terria.workbench.contains(memberModel);
        if (allItemsLoaded) {
          addOrRemoveMember(memberModel, viewState, keepCatalogOpen);
        } else if (!allItemsLoaded && !itemLoaded) {
          addOrRemoveMember(memberModel, viewState, keepCatalogOpen);
        }
      }
    });
```

In this way the logic achieves what we want, without writing a duplicate function with similar role, and without introduing more parameters like `forceState`.

However, I get your point that rewriting it could make it clearer? What do you think? I think given the comments in the code its clear enough personally?

Reply:
Got it.

I'll probably also leave a comment inside the branches of the `if` condition just to be extra clear.

Reply:
OK will do!